### PR TITLE
implement promote for BitArray

### LINKED ISSUE #43551: promote fails with BitArrays
```
promote(randn(4,5), bitrand(2,2))
```

> ERROR: promotion of types Matrix{Float64} and BitMatrix failed to change any arguments
> Stacktrace:
>  [1] error(::String, ::String, ::String)
>    @ Base ./error.jl:42
>  [2] sametype_error(input::Tuple{Matrix{Float64}, BitMatrix})
>    @ Base ./promotion.jl:374
>  [3] not_sametype(x::Tuple{Matrix{Float64}, BitMatrix}, y::Tuple{Matrix{Float64}, BitMatrix})
>    @ Base ./promotion.jl:368
>  [4] promote(x::Matrix{Float64}, y::BitMatrix)
>    @ Base ./promotion.jl:351
>  [5] top-level scope
>    @ REPL[5]:1
> 

Contrast that with,

```
julia> promote(randn(4,5), rand(-3:3, 2,2))

([1.0404679330906446 0.13605200544037485 … 0.6584649350909583 1.759298096581398; 0.8577584331870316 -0.9385873914104255 … -0.7270472742833658 0.42998694669894727; -1.3226103979280828 1.2291060826052171 … 2.4340481349552263 -0.3980550678017092; 0.3515606302511447 -0.2598112417308294 … -0.2447667146035612 -1.6238446588858528], [2.0 2.0; 0.0 0.0])
```

### PR DESCRIPTION
**Pull request by vtjnash**
Fix #43551
Replaces #43646 @[JakobSachs](https://github.com/JakobSachs) following my comment there, based on your work

### CODE REVIEW

**Comment by vtjnash** [comment]
@JakobSachs this turned out to be more complicated than I thought. WDYT?

@JeffBezanson Thought you might be interested to look at the new `promote_rule`/`promote_type` fallback behavior added here.

File: base/array.jl [hunk 1]
```diff
@@ -617,7 +617,7 @@ oneunit(x::AbstractMatrix{T}) where {T} = _one(oneunit(T), x)
 convert(::Type{T}, a::AbstractArray) where {T<:Array} = a isa T ? a : T(a)
 convert(::Type{Union{}}, a::AbstractArray) = throw(MethodError(convert, (Union{}, a)))
 
-promote_rule(a::Type{Array{T,n}}, b::Type{Array{S,n}}) where {T,n,S} = el_same(promote_type(T,S), a, b)
+promote_rule(a::Type{<:Array{T,n}}, b::Type{<:Array{S,n}}) where {T,n,S} = el_same(promote_type(T, S), a, b)
```
Comment (vtjnash) [inline base/array.jl:620]:
```suggestion
promote_rule(a::Type{Array{T,n}}, b::Type{Array{S,n}}) where {T,n,S} = el_same(promote_type(T, S), a, b)
```

**Comment by JakobSachs** [comment]
Seems good to me, tho i cant say i full understand the changes you made, regarding the indices.

File: base/range.jl [hunk 2]
```diff
@@ -1263,13 +1263,17 @@ function -(r::LinRange)
...
-el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{T,n}}) where {T,n}   = a
+el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{T,n}}) where {T,n}   = a # we assume a === b
 el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{S,n}}) where {T,S,n} = a
 el_same(::Type{T}, a::Type{<:AbstractArray{S,n}}, b::Type{<:AbstractArray{T,n}}) where {T,S,n} = b
 el_same(::Type, a, b) = promote_typejoin(a, b)
 
+promote_result(::Type{<:AbstractArray}, ::Type{<:AbstractArray}, ::Type{T}, ::Type{S}) where {T,S} = (@inline; promote_type(T,S))
+promote_result(::Type{T}, ::Type{S}, ::Type{Bottom}, ::Type{Bottom}) where {T<:AbstractArray,S<:AbstractArray} = (@inline; promote_typejoin(T,S))
+# If no promote_rule is defined, both directions give Bottom. In that case use typejoin on the eltypes instead and give Array as the container.
+promote_result(::Type{<:AbstractArray{T,n}}, ::Type{<:AbstractArray{S,n}}, ::Type{Bottom}, ::Type{Bottom}) where {T,S,n} = (@inline; Array{promote_type(T,S),n})
```
Comment (N5N3) [inline base/range.jl:1275]: @vtjnash Line 1275 breaks `StaticArrays.jl`'s test on master.
After this PR
```julia
julia> [(@MArray [1;2;3]), (@SArray [1;2;3])]
2-element Vector{Vector{Int64}}:
 [1, 2, 3]
 [1, 2, 3]
```
It was a `Vector{Any}` before this PR.
`for a in [array1, array2]` is general pattern for test. So this change seems a little "break"?

Comment (vtjnash) [reply base/range.jl:1275]: Test should not do that, that is an incorrect test. `[a, b]` is a general pattern for changing `a` and `b` to a common type. If you do not want that, you must specify the type of the array you want (usually `Any[...]`)

Comment (nalimilan) [reply base/range.jl:1275]: FWIW CategoricalArrays tests also used this pattern. Actually this revealed the lack of `promote_type` methods for `CategoricalArray`s.

This has the potential to disrupt lots of tests, as things like e.g. `[["b", "a"], ["b", missing]]` will suddenly give two `Vector{Union{Missing, String}}` objects while previously the first one was `Vector{String}`. I'm not saying it shouldn't be done though but we have to be ready to notice some breakage.

Comment (nalimilan) [inline base/range.jl:1275]: Why define these in range.jl? When I was looking for such methods I looked in promotion.jl and in abstractarray.jl and I couldn't find them.

**Pull request merged**